### ISSUE TITLE
Move routing.map_http_methods into an instance or class method?

### ISSUE THREAD

**Issue by xgfone** [falconry/falcon#1250]
I like `Falcon`, but the policy to map http methods. So I want to customize it, and I must override the method `add_route`, which is `WASTEFUL` and `EXPENSIVE`.

I think, `method_map = routing.map_http_methods(resource, suffix=suffix)` should be placed in an instance or class method, such as
```python
class API(object):
    def map_http_methods(self, resource, *args, **kwargs)
        return routing.map_http_methods(resource, suffix=kwargs.pop('suffix', None))

    def add_route(uri_template, resource, *args, **kwargs):
        # ...
        method_map = self.map_http_methods(resource, *args, **kwargs)
        routing.set_default_responders(method_map)
        self._router.add_route(uri_template, method_map, resource, *args, **kwargs)
```

Now, I only need to override the method `map_http_methods`, don't change or repeat the process in `add_route`. For example,
```python
class MyAPI(API):
    def map_http_methods(self, resource, *args, **kwargs):
        return {
            kwargs["method"]: getattr(resource, kwargs["action"])
        }

        # OR

        return {method: getattr(resource, action) for method, action in kwargs["map"].items()}
```

I know someone does not like this method mapping, but you should not forbid others to like it.

**Comment by xgfone** [comment]
@kgriffs I have submitted a RP #1252 , please take a look. Thanks.

This is my overriding example,
```python
class MyAPI(falcon.API):
    
    def map_http_methods(self, resource, args, kwargs):
        _map = kwargs.pop("map", None)
        if _map:
            return {k.upper(): getattr(resource, v) for k, v in _map.items()}
        return super(MyAPI, self).map_http_methods(resource, args, kwargs)
```

It supports not only the original HTTP method mapping but also the new self-defined. For example,
```python
resource = Resource()
app = MyAPI()
app.add_route("/path/to/1", resource)  # The original
app.add_route("/path/to/2", resource, map={"GET": "my_get_method", "POST": "my_post_method"})
```

@kgriffs I'm thinking whether we maybe use the following method instead of that in PR.
```python
def map_http_methods(self, resource, args, kwargs):
    _map = kwargs.pop("map", None)
    if _map:
        return {k.upper(): getattr(resource, v) for k, v in _map.items()}
        
    suffix = kwargs.pop('suffix', None)
    return routing.map_http_methods(resource, suffix=suffix)
```

**Comment by kgriffs** [comment]
Hi @xgfone, it probably makes sense to be somewhat conservative for now. We can wait and see if there are common use cases for which the new method is overridden, and then perhaps implement those in the framework over time. I'm looking at the PR now. Thanks for your help on this!

**Comment by kgriffs** [comment]
BTW, could you explain in a bit more detail how you are setting up your resource classes and how you need the routing to work? I just want to make sure I understand the use case.

**Comment by xgfone** [comment]
Hi @kgriffs, I have rebased and modified it based on the new interface.

I mainly have two ideas,
1. Add the method `map_http_methods` to allow the user to override it to implement the self-defined method mapping.
2. The original function maps HTTP methods to methods `on_XXX()` of the resource. I think, in order to make the framework more flexible, it maybe allow the user to appoint the method name of the resource, not `on_XXX()`, such as `app.add_route("/path/to", resource, map={"GET": "resource_method"})` or `app.add_route("/path/to", resource, get="resource_method")`.

Now, I have only implemented the first in order to keep the consistent with the original.

**Comment by xgfone** [comment]
@kgriffs Just as you said,  it is best to make the new, override-able method a member of the router rather than `falcon.API`. Mapping the HTTP method should be the responsibility of the router, not `falcon.API`. 

`CompiledRouter.add_route()` will be broken down, because the argument `method_map` is unnecessary.

I have done. Please took a look, and I need your thoughts.

**Comment by vytas7** [comment]
Hi @xgfone ,
Does the PR https://github.com/falconry/falcon/pull/1329 address your concerns, or is there still work left according to your proposal?

Thinking if we could close this issue now that https://github.com/falconry/falcon/pull/1329 is merged.

**Comment by vytas7** [comment]
Hi again @xgfone !
Closing this issue since the main problem (custom routers needing to reimplement method mapping) seems to be addressed, and well tested as part of Falcon 2.0 released last year.
